CSharpSourceCode/Battle/BlackPowderWeapon/BlackPowderWeaponMissionLogic.cs

```diff
@@ -55,41 +65,66 @@ public override void OnAgentShootMissile(Agent shooterAgent, EquipmentIndex weap
+                }
+            }
+        }
+
+        /// <summary>
+        /// The script will check each ammo weapon (that fits current usage of shooter's weapon) for bullets to do blunderbuss shot. 
+        /// It can use multiple ammo weapons to collect required amount of bullets.
+        /// </summary>
+        /// <param name="requiredAmmoAmount">Max amount of bullets to do blunderbuss shot.</param>
+        private void DoBlunderbussShot(Agent shooterAgent, Vec3 shotPosition, Mat3 shotOrientation, short requiredAmmoAmount)
+        {
+            MissionWeapon weapon = MissionWeapon.Invalid;
+            short foundAmmoAmount = 0;
+            for (EquipmentIndex index = EquipmentIndex.WeaponItemBeginSlot; index < EquipmentIndex.NumAllWeaponSlots; index++)
+            {
+                weapon = shooterAgent.Equipment[index];
+                // Weapon hit points mean amount of ammo
+                if (weapon.CurrentUsageItem.WeaponClass == shooterAgent.WieldedWeapon.CurrentUsageItem.AmmoClass && weapon.HitPoints > 0)
                 {
-                    if (shooterAgent.WieldedWeapon.Item.StringId.Contains("two_barrels"))
+                    foundAmmoAmount += Math.Min(requiredAmmoAmount, weapon.HitPoints);
+                    short newAmount = (short)(weapon.HitPoints - Math.Min(foundAmmoAmount, weapon.HitPoints));
+                    shooterAgent.SetWeaponAmountInSlot(index, newAmount, false);
+                    if (foundAmmoAmount == requiredAmmoAmount)
                     {
-                        DoTwoBarrelsShot(shooterAgent, position, orientation);
+                        break;
                     }
-                    else if (shooterAgent.WieldedWeapon.Item.StringId.Contains("four_barrels"))
+                    else
                     {
-                        DoFourBarrelsShot(shooterAgent, position, orientation);
+                        continue;
                     }
                 }
             }
-        }
 
-        private void DoBlunderbussShot(Agent shooterAgent, Vec3 position, Mat3 orientation)
-        {
-            var weaponData = shooterAgent.WieldedWeapon.CurrentUsageItem;
-            var scattering = 1f / (weaponData.Accuracy * 1.2f);
-            for (int i = 0; i < 10; i++)
+            WeaponComponentData weaponData = shooterAgent.WieldedWeapon.CurrentUsageItem;
+            float scattering = 1f / (weaponData.Accuracy * 1.2f);
+            while (foundAmmoAmount > 0)
             {
-                var missile = shooterAgent.WieldedWeapon.AmmoWeapon;
-                var _orientation = GetRandomOrientationForBlunderbass(orientation, scattering);
-                Mission.AddCustomMissile(shooterAgent, missile, position, _orientation.f, _orientation, weaponData.MissileSpeed, weaponData.MissileSpeed, false, null);
+                foundAmmoAmount--;
+                var _orientation = GetRandomOrientationForBlunderbass(shotOrientation, scattering);
```

Comment:
If I am honest, the naming could be better. you calculate a spreading, and it would be better if this is somehow contained here. I am also confident, that a spreading could be later used for different weapons, that are not a Blunderbass.

Reply:
yeah, you're right, fixed

Reply:
i also created BlackPowderWeaponScript abstract class for exploding missiles and reworked grenade methods. Now there is only one AddMissileScript method